src/stub/preprocessor/forward_declarations.rs

```diff
@@ -26,53 +26,75 @@ struct MainWrapper {
+///   variable, includes all the fields in a VariableCommand (not nested under
+///   `var`)
+/// - `main_wrapper` - wraps all of the code, contains `forward_declarations`
+///   (the above resource, rendered) and `main_content`
 pub fn transform(stub: &mut Stub) {
     let mut max_nested_depth = 0;
 
-    let mut forward_declarations: Vec<VariableCommand> = stub.commands.iter().filter_map(|cmd| {
+    let mut forward_declarations: Vec<VariableCommand> = stub
+        .commands
+        .iter()
+        .filter_map(|cmd| {
         let (cmd, nested_depth) = unpack_cmd(cmd, 0);
 
         if nested_depth > max_nested_depth {
             max_nested_depth = nested_depth;
         }
 
-        if let Cmd::Read(vars) = cmd {
-            Some(vars.into_iter())
+            if let Cmd::Read(var_cmds)
+            | Cmd::LoopLine {
+                variables: var_cmds, ..
+            } = cmd
+            {
+                Some(var_cmds.into_iter())
         } else {
             None
         }
-    }).flatten().collect();
+        })
+        .flatten()
+        .collect();
 
-    let mut loop_vars: Vec<VariableCommand> = ALPHABET[0..max_nested_depth].iter().filter(|loop_var|
-        forward_declarations.iter().all(|var_cmd| var_cmd.ident != loop_var.to_string())
-    ).map(|loop_var|
-        VariableCommand { 
+    let mut loop_vars: Vec<VariableCommand> = ALPHABET[0..max_nested_depth - 1]
+        .iter()
+        .filter(|loop_var| forward_declarations.iter().all(|var_cmd| var_cmd.ident != loop_var.to_string()))
+        .map(|loop_var| VariableCommand {
             ident: loop_var.to_string(), 
             var_type: VarType::Int, 
             max_length: None, 
-            input_comment: String::new() 
-        }
-    ).collect();
+            input_comment: String::new(),
+        })
+        .collect();
 
     forward_declarations.append(&mut loop_vars);
 
+    let mut unique_forward_declarations: Vec<VariableCommand> = Vec::new();
+    let mut seen_idents: Vec<String> = Vec::new();
+
+    for var_cmd in &forward_declarations {
+        if !seen_idents.contains(&var_cmd.ident) {
+            seen_idents.push(var_cmd.ident.clone());
+            unique_forward_declarations.push(var_cmd.clone());
+        }
+    }
+
     let wrapper = MainWrapper {
-        forward_declarations,
+        forward_declarations: unique_forward_declarations,
         main_content: stub.commands.drain(..).collect(),
     };
 
     stub.commands = vec![Cmd::External(Box::new(wrapper))];
 }
 
 fn unpack_cmd(cmd: &Cmd, nested_depth: usize) -> (Cmd, usize) {
-    if let Cmd::Loop { count_var: _, command: subcmd } = cmd {
-        unpack_cmd(subcmd, nested_depth + 1)
-    } else {
-        (cmd.clone(), nested_depth + 1)
+    match cmd {
+        Cmd::Loop {
+            count_var: _,
+            command: subcmd,
+        } => unpack_cmd(subcmd, nested_depth + 1),
+        _ => (cmd.clone(), nested_depth + 1),
```

Comment:
```suggestion
        _ => (cmd.clone(), nested_depth)
```

Instead of subtracting `1` from max_nested_depth above.

Reply:
Changed.